### PR TITLE
Qrack Plugin

### PR DESCRIPTION
**Pull request by WrathfulSpatula**
This branch adds a plugin for the vm6502q/qrack library. Qrack provides optional OpenCL-based GPU acceleration, as well as a novel optimization layer including "Schmidt decomposition," as suggested by Pednault and team in [Breaking the 49-Qubit Barrier in the Simulation of Quantum Circuits](https://arxiv.org/abs/1710.05867). (The authors of Qrack are not in any way affiliated with the authors of that paper, but we thank them.) This plugin was adapted from XACC's Q++ plugin.

Thank you folks!

### CODE REVIEW

**Comment by WrathfulSpatula** [comment]
Hate to slide this in on top, but it occurred to me how to fix an issue: I was having difficulty fixing the warning for undefined OpenCL version macros in each linked executable. Eventually, it felt like defeat, but I defined the macros in the QrackVisitor header instead of in CMake. I was worried this might conflict with other linked code that defines different version metadata, (there's no reason Qrack shouldn't work with OpenCL 1.2 or 3.0, in addition to 2.0, as far as I know,) but it should be fine, if we just wrap these macros in `#ifndef`.

I can squash this again, if anyone would like me to. Thanks again!

**Comment by amccaskey** [comment]
@tnguyen-ornl would you also take a quick look at this and see if I missed anything?

File: quantum/plugins/qrack/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+usfunctiongetresourcesource(TARGET ${LIBRARY_NAME} OUT SRC)
+usfunctiongeneratebundleinit(TARGET ${LIBRARY_NAME} OUT SRC)
+
+add_library(${LIBRARY_NAME} SHARED ${SRC})
+
+target_include_directories(${LIBRARY_NAME}
+                             PUBLIC .
+                                    ./accelerator
+                                    ./accelerator/src/include
+                                    ${CMAKE_SOURCE_DIR}/tpls/eigen)
```
Comment (tnguyen-ornl) [inline quantum/plugins/qrack/CMakeLists.txt:16]: Are you using `eigen` in qrack?

**Comment by tnguyen-ornl** [comment]
Lgtm! It's pretty well-written and easy to follow :)

Since this accelerator has some interesting special use cases and params, perhaps it's worthwhile adding a few notes to https://xacc.readthedocs.io/en/latest/extensions.html#accelerators.

File: quantum/plugins/qrack/CMakeLists.txt (hunk 1 above)
Comment (WrathfulSpatula) [reply quantum/plugins/qrack/CMakeLists.txt:16]: No, I am not. This was left behind in migrating from the Quantum++ plugin, and I will remove it.

Comment (WrathfulSpatula) [reply quantum/plugins/qrack/CMakeLists.txt]: Done.

**Comment by WrathfulSpatula** [comment]
@amccaskey 

- QrackAccelerator.hpp and QrackVisitor.hpp, should be pragma once - Fixed.
- QrackAccelerator initialize line 35, the key was wrong - Fixed.

@tnguyen-ornl  I will open a separate pull on the docs repo, for the Qrack plugin parameters. (The constructor arguments are also documented in the Qrack docs, but they bear repetition here for ease and clarity.)

**Pull request merged**